### PR TITLE
Error refactoring

### PR DESCRIPTION
**Pull request by Nerzal**
Wrap Errors and use APIError struct to also provide the httpstatus code.

Adresses: #132

### CODE REVIEW

**Comment by Nerzal** [comment]
gnah xD

**Comment by Nerzal** [comment]
this will most likely result in some merges as soon as i'm finished xD working through 1,4k lines of code :D

**Comment by SVilgelm** [comment]
Could be, just restore go.mod and go.sum

**Comment by Nerzal** [comment]
I'm not quite sure, if we also pack this 

```go
	if resp.StatusCode() == 409 {
			return &ObjectAlreadyExists{ErrorMessage: msg}
	}
```

into an APIError as it feels like a specific error is better to handle :thinking:

**Comment by SVilgelm** [comment]
> into an APIError as it feels like a specific error is better to handle 🤔

I think we should always return ApiError, we can easy check if status == 409

**Comment by Nerzal** [comment]
atleast we should treat these changes as breaking changes, when going up with the version, as the some integrations might check for specific error messages and/or the now removed error types.

**Comment by Nerzal** [comment]
i'll try to fix the failing builds later

**Comment by SVilgelm** [comment]
> atleast we should treat these changes as breaking changes
sure

File: client_test.go [hunk 1]
```diff
@@ -305,7 +305,9 @@ func SetUpTestUser(t testing.TB, client GoCloak) {
 			cfg.GoCloak.Realm,
 			user,
 		)
-		if IsObjectAlreadyExists(err) {
+
+		apiError := err.(*APIError)
+		if apiError.Code == http.StatusConflict {
```
Comment (SVilgelm) [inline client_test.go]: 👍

**Comment by SVilgelm** [comment]
Ok, I see. Please do something with:
`actual  : "could not get client scope: Code: 404 Message: 404 Not Found: Could not find client scope"`
I think it's not good idea to repeat 404 several times

**Comment by Nerzal** [comment]
true, i'll have a look into it

**Comment by Nerzal** [comment]
finally (:

**Comment by SVilgelm** [comment]
v5 released!!!!

**Comment by SVilgelm** [comment]
oh no

**Comment by SVilgelm** [comment]
need to remove v5

**Pull request merged**